File: macos/Tests/SnapBriefCoreTests/ExtendedCommentsTests.swift
// Port of `tests/SnapBrief.Core.Tests/ExtendedCommentsTests.cs`, SPEC-DELTA-2.md §5,
// SPEC-DELTA-2B.md §B (linked comments beyond one alphabet).
import XCTest

@testable import SnapBriefCore

final class ExtendedCommentsTests: XCTestCase {
    func test_Labels_extend_beyond_one_alphabet() throws {
        XCTAssertEqual(try CaptureLabels.forIndex(0), "A")
        XCTAssertEqual(try CaptureLabels.forIndex(25), "Z")
        XCTAssertEqual(try CaptureLabels.forIndex(26), "AA")
        XCTAssertEqual(try CaptureLabels.forIndex(299), "KN")
        XCTAssertThrowsError(try CaptureLabels.forIndex(-1))
    }

    func test_Hundreds_of_comments_preserve_identity_links_and_arrow_style_in_json() throws {
        let parent = AnnotationItem.create(
            kind: .rectangle, points: [NormalizedPoint(0.05, 0.05), NormalizedPoint(0.2, 0.2)], note: "Родитель")

        var comments: [AnnotationItem] = []
        for index in 0..<300 {
            comments.append(
                AnnotationItem.create(
                    kind: .comment,
                    points: [NormalizedPoint(0.5, 0.5), NormalizedPoint(0.51, 0.51)],
                    note: "Комментарий \(index)",
                    parentAnnotationId: parent.id,
                    arrowStyle: "curved"))
        }

        var capture = CaptureItem.create(sourceImagePath: "source/a.png", pixelWidth: 1000, pixelHeight: 800)
        capture.annotations = [parent] + comments

        var session = SnapBriefSession.create(nowUtc: Date())
        session.captures = [capture]

        let data = try SnapBriefJson.encoder.encode(session)
        let decoded = try SnapBriefJson.decoder.decode(SnapBriefSession.self, from: data)

        XCTAssertEqual(decoded.captures.count, 1)
        let decodedAnnotations = decoded.captures[0].annotations
        XCTAssertEqual(decodedAnnotations.count, 301)
        let decodedComments = decodedAnnotations.filter { $0.kind == .comment }
        XCTAssertEqual(decodedComments.count, 300)
        XCTAssertTrue(decodedComments.allSatisfy { $0.parentAnnotationId == parent.id })
        XCTAssertTrue(decodedComments.allSatisfy { $0.arrowStyle == "curved" })

        let prompt = try PromptGenerator().generate(decoded)
        XCTAssertTrue(prompt.contains("A301: Комментарий 299 (к области A1)"))
    }

    func test_Cropping_keeps_comment_but_clears_a_removed_parent_link() throws {
        let parent = AnnotationItem.create(
            kind: .rectangle, points: [NormalizedPoint(0.6, 0.6), NormalizedPoint(0.9, 0.9)], note: "Parent")
        let comment = AnnotationItem.create(
            kind: .comment, points: [NormalizedPoint(0.1, 0.1), NormalizedPoint(0.11, 0.11)], note: "Keep",
            parentAnnotationId: parent.id, arrowStyle: "straight")

        var source = CaptureItem.create(sourceImagePath: "source/original.png", pixelWidth: 1000, pixelHeight: 800)
        source.annotations = [parent, comment]

        let result = try CaptureCropper.crop(
            source: source,
            cropBounds: NormalizedRect(0, 0, 0.5, 0.5),
            croppedSourceImagePath: "source/cropped.png",
            croppedPixelWidth: 500,
            croppedPixelHeight: 400)

        XCTAssertEqual(result.removedAnnotationIds, [parent.id])
        XCTAssertEqual(result.croppedCapture.annotations.count, 1)
        let retainedComment = try XCTUnwrap(result.croppedCapture.annotations.first)
        XCTAssertEqual(retainedComment.id, comment.id)
        XCTAssertEqual(retainedComment.note, "Keep")
        XCTAssertNil(retainedComment.parentAnnotationId)
    }
}

File: tests/SnapBrief.Core.Tests/ExtendedCommentsTests.cs
using System.Collections.Immutable;
using SnapBrief.Core.Exporting;
using SnapBrief.Core.Models;
using SnapBrief.Infrastructure.Serialization;

namespace SnapBrief.Core.Tests;

public sealed class ExtendedCommentsTests
{
    [Theory]
    [InlineData(0, "A")]
    [InlineData(25, "Z")]
    [InlineData(26, "AA")]
    [InlineData(299, "KN")]
    public void Labels_extend_beyond_one_alphabet(int index, string expected) => Assert.Equal(expected, CaptureLabels.ForIndex(index));

    [Fact]
    public void Hundreds_of_comments_preserve_identity_links_and_arrow_style_in_json()
    {
        var parent = AnnotationItem.Create(AnnotationKind.Arrow, [new(.1, .1), new(.5, .5)], note: "Область") with { ArrowStyle = "curved" };
        var notes = Enumerable.Range(0, 300).Select(i => AnnotationItem.Create(AnnotationKind.Comment,
            [new(.2, .2), new(.21, .21)], note: $"Комментарий {i}") with { ParentAnnotationId = parent.Id }).ToImmutableArray();
        var capture = CaptureItem.Create("source/a.png", 1000, 1000) with { Annotations = notes.Insert(0, parent) };
        var session = SnapBriefSession.Create(DateTimeOffset.UtcNow) with { Captures = [capture] };
        var json = System.Text.Json.JsonSerializer.Serialize(session, SnapBriefJson.Options);
        var restored = System.Text.Json.JsonSerializer.Deserialize<SnapBriefSession>(json, SnapBriefJson.Options)!;
        Assert.Equal("curved", restored.Captures[0].Annotations[0].ArrowStyle);
        Assert.All(restored.Captures[0].Annotations.Skip(1), a => Assert.Equal(parent.Id, a.ParentAnnotationId));
        var prompt = new PromptGenerator().Generate(restored);
        Assert.Contains("A301: Комментарий 299 (к области A1)", prompt);
    }

    [Fact]
    public void Note_offset_survives_json_and_stays_null_in_a_session_written_without_it()
    {
        var moved = AnnotationItem.Create(AnnotationKind.Comment, [new(.2, .2), new(.21, .21)], note: "Сдвинутая")
            with { NoteOffset = new NormalizedPoint(-.08, .15) };
        var capture = CaptureItem.Create("source/a.png", 1000, 1000) with { Annotations = [moved] };
        var session = SnapBriefSession.Create(DateTimeOffset.UtcNow) with { Captures = [capture] };
        var json = System.Text.Json.JsonSerializer.Serialize(session, SnapBriefJson.Options);
        var restored = System.Text.Json.JsonSerializer.Deserialize<SnapBriefSession>(json, SnapBriefJson.Options)!;
        Assert.Equal(new NormalizedPoint(-.08, .15), restored.Captures[0].Annotations[0].NoteOffset);

        var legacy = System.Text.Json.Nodes.JsonNode.Parse(json)!;
        legacy["captures"]![0]!["annotations"]![0]!.AsObject().Remove("noteOffset");
        var beforeTheField = System.Text.Json.JsonSerializer.Deserialize<SnapBriefSession>(legacy.ToJsonString(), SnapBriefJson.Options)!;
        Assert.Null(beforeTheField.Captures[0].Annotations[0].NoteOffset);
        SessionValidation.Validate(beforeTheField);
        SessionValidation.Validate(restored);
    }

    [Fact]
    public void Shape_and_fill_survive_json_and_default_for_a_session_written_without_them()
    {
        var oval = AnnotationItem.Create(AnnotationKind.Rectangle, [new(.1, .1), new(.6, .4)])
            with { Shape = AnnotationShape.Ellipse, Fill = AnnotationFill.Translucent };
        var capture = CaptureItem.Create("source/a.png", 1000, 1000) with { Annotations = [oval] };
        var session = SnapBriefSession.Create(DateTimeOffset.UtcNow) with { Captures = [capture] };
        var json = System.Text.Json.JsonSerializer.Serialize(session, SnapBriefJson.Options);
        Assert.Contains("\"shape\": \"ellipse\"", json);
        Assert.Contains("\"fill\": \"translucent\"", json);
        var restored = System.Text.Json.JsonSerializer.Deserialize<SnapBriefSession>(json, SnapBriefJson.Options)!;
        Assert.Equal(AnnotationShape.Ellipse, restored.Captures[0].Annotations[0].Shape);
        Assert.Equal(AnnotationFill.Translucent, restored.Captures[0].Annotations[0].Fill);

        var legacy = System.Text.Json.Nodes.JsonNode.Parse(json)!;
        var written = legacy["captures"]![0]!["annotations"]![0]!.AsObject();
        written.Remove("shape");
        written.Remove("fill");
        var beforeTheFields = System.Text.Json.JsonSerializer.Deserialize<SnapBriefSession>(legacy.ToJsonString(), SnapBriefJson.Options)!;
        Assert.Equal(AnnotationShape.Rectangle, beforeTheFields.Captures[0].Annotations[0].Shape);
        Assert.Equal(AnnotationFill.None, beforeTheFields.Captures[0].Annotations[0].Fill);
        SessionValidation.Validate(beforeTheFields);
    }

    // The preview window checked these two rules on its own list of comments. The window is gone,
    // so the rules are checked here, on the labels the export itself hands out.
    [Fact]
    public void An_empty_note_claims_no_label_and_numbering_closes_the_gap_after_a_deletion()
    {
        var first = AnnotationItem.Create(AnnotationKind.Rectangle, [new(.1, .1), new(.2, .2)], note: "Первый");
        var blank = AnnotationItem.Create(AnnotationKind.Comment, [new(.3, .3), new(.31, .31)]);
        var second = AnnotationItem.Create(AnnotationKind.Comment, [new(.4, .4), new(.41, .41)], note: "Второй");
        var capture = CaptureItem.Create("source/a.png", 1000, 1000) with { Annotations = [first, blank, second] };

        var labels = CaptureLabels.ForNotedAnnotations("A", capture).ToArray();
        Assert.Equal(["A1", "A2"], labels.Select(item => item.DisplayLabel));
        Assert.DoesNotContain(blank.Id, labels.Select(item => item.Annotation.Id));

        var afterDeletion = capture with { Annotations = capture.Annotations.Remove(first) };
        var remaining = CaptureLabels.ForNotedAnnotations("A", afterDeletion).ToArray();
        Assert.Equal(["A1"], remaining.Select(item => item.DisplayLabel));
        Assert.Equal(second.Id, remaining.Single().Annotation.Id);
    }

    [Fact]
    public void Cropping_keeps_comment_but_clears_a_removed_parent_link()
    {
        var parent = AnnotationItem.Create(AnnotationKind.Rectangle, [new(.8, .8), new(.9, .9)]);
        var comment = AnnotationItem.Create(AnnotationKind.Comment, [new(.1, .1), new(.11, .11)], note: "Keep") with { ParentAnnotationId = parent.Id };
        var capture = CaptureItem.Create("source/a.png", 1000, 1000) with { Annotations = [parent, comment] };
        var cropped = SnapBrief.Core.Editing.CaptureCropper.Crop(capture, new(0, 0, .5, .5), "source/crop.png", 500, 500).CroppedCapture;
        Assert.Equal(comment.Id, Assert.Single(cropped.Annotations).Id);
        Assert.Null(cropped.Annotations[0].ParentAnnotationId);
        Assert.Equal("Keep", cropped.Annotations[0].Note);
    }
}

